src/js/server/kettleValidation.js

```diff
@@ -0,0 +1,107 @@
+var gpii  = fluid.registerNamespace("gpii");
+
+fluid.registerNamespace("gpii.schema.kettle.validator");
+
+gpii.schema.kettle.validator.validateRequest = function (kettleValidator, globalValidator, requestHandler) {
+    var schemaKey = requestHandler.options.schemaKey || "default";
+    var rulesKey  = requestHandler.options.rulesKey  || "default";
+    var schemaHash = kettleValidator.schemaHashes[schemaKey];
+    var gssSchema = fluid.get(kettleValidator, ["options", "requestSchemas", schemaKey]);
+    var transformationRules = fluid.get(kettleValidator, ["options", "requestContentToValidate", rulesKey]);
+    var toValidate = fluid.model.transformWithRules(requestHandler.req, transformationRules);
+    var validationResults = globalValidator.validate(gssSchema, toValidate, schemaHash);
+
+    if (validationResults.isValid) {
+        requestHandler.handleValidRequest(requestHandler);
+    }
+    else {
+        var localisedErrors = gpii.schema.validator.localiseErrors(validationResults.errors, toValidate, kettleValidator.model.messages, kettleValidator.options.localisationTransform);
+        var localisedPayload = fluid.copy(validationResults);
+        localisedPayload.errors = localisedErrors;
+
+        var failurePayload = fluid.extend({}, kettleValidator.options.errorTemplate, localisedPayload);
+        requestHandler.events.onError.fire(failurePayload);
+    }
+};
+
+gpii.schema.kettle.validator.cacheSchemas = function (kettleValidator, globalValidator) {
+    fluid.each(kettleValidator.options.requestSchemas, function (requestSchema, schemaKey) {
+        var schemaHash = gpii.schema.hashSchema(requestSchema);
+        kettleValidator.schemaHashes[schemaKey] = schemaHash;
+        globalValidator.cacheSchema(requestSchema, schemaHash);
+    });
+};
+
+fluid.defaults("gpii.schema.kettle.validator", {
+    gradeNames: ["fluid.modelComponent"],
+    members: {
+        schemaHashes: {}
+    },
+    model: {
+        messages: gpii.schema.messages.validationErrors
+    },
+    localisationTransform: {
+        "": ""
+    },
+    errorTemplate: {
+        // "Bad Request": https://developer.mozilla.org/nl/docs/Web/HTTP/Status/400
+        statusCode: 400,
+        message: "Your request was invalid.  See the errors for details."
+    },
+    requestSchemas: {
+        default: {
+            "$schema": "gss-v7-full#"
+        }
+    },
+    requestContentToValidate: {
+        default: "{that}.options.requestContentToValidate.body",
+        body: {
+            "": "body"
+        },
+        params: {
+            "": "params"
+        },
+        query: {
+            "": "query"
+        }
+    },
+    invokers: {
+        validateRequest: {
+            funcName: "gpii.schema.kettle.validator.validateRequest",
+            args: ["{that}", "{gpii.schema.validator}", "{arguments}.0"] // kettleValidator, globalValidator, request
+        }
+    },
+    listeners: {
+        "onCreate.cacheSchemas": {
```

Comment:
It will make life easier if, rather than one validator per app, we instead have one per request type. Then we can either have a middleware component as a single static child, or depending on how messy it is, have the validator be the middleware itself.

Reply:
You could, of course, carry on hosting all validators in a single static component in this way, but this then raises the problem of how you arrange for all the middleware components to be constructed sufficiently early - this could be done but would probably involve something messy such as a synthesised raw dynamic grade definition.